core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java

```diff
@@ -20,9 +20,19 @@
 package org.elasticsearch.common.inject;
 
 /**
+ * This interface can be added to a Module to spawn sub modules. DO NOT USE.
  *
+ * This is fundamentally broken.
+ * <ul>
+ * <li>If you have a plugin with multiple modules, return all the modules at once.</li>
+ * <li>If you are trying to make the implementation of a module "pluggable", don't do it.
+ * This is not extendable because custom implementations (using onModule) cannot be
+ * registered before spawnModules() is called.</li>
+ * </ul>
+ * @deprecated Create modules directly
  */
-public interface SpawnModules {
+@Deprecated
```

Comment:
> (1) more classes

sure, I can leave with a proposed alternative that will remove the class, but quite frankly, I don't care if we have an extra class here

> (2) more obfuscation

note sure what's obfuscated here... it's as simple as "modules that implement this depend on other modules"

>  (3) possibility for broken pluggability APIs

Give me any extensible module that doesn't implement this interface and I'll show you how I can implement it in a broken way. As I mentioned, this interface is not for extensibility, but for modularity. Abusing things can be done everywhere... and we need to make sure that we carefully define the contract for APIs. And documentation **is** part of the contract - I'll be happy to see an explanation in the javadoc that states 1) what this interface is for and all about in general, 2) that you should not abuse it for extensibility and if you do want extensibility point to a top package level javadoc that explains what is the right way of opening an extension point.

> I don't know what you mean by "module pre processing"

https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/common/inject/PreProcessModule.java

> but extension points should be clear, and plugins should bind ancillary classes they need in their own modules.

totally agree, and that's also the case with the `SpanwModules` itself. The only thing that happens here in addition is the ability to group modules per functionality/layer if you will. To your example, if I'm replacing the `TransportModule`, I don't want to declare that I'm replacing it in one place and in another place define the sub-modules of the new transport module. I want to simply declare "this is the new transport module" and it comes with whatever sub-modules that it defines - one place - clean. On the same note, if you choose to remove/replace a module, I don't want to think carefully what other modules I need to replace along with it - that what you'll end up with if you flatten all the modules - there're no clear module boundaries and dependencies. 

The problem that you're referring to in the javadoc (and I generally agree that there is a problem here) should be fixed, but in the appropriate place and the spawn modules should not be "blamed" for it. The problem here is that there needs to be a well defined function for replacing modules, on the container level.. somewhere you can simply call:

```
replace(TransportModule.class).with(MyTransportModule.class)
```

and the implementation should take care of unregistering the sub-modules of the prev. one and registering the sub-modules of the new one. 

Removing this interface will prevent defining well define & separated modules and promote module spaghetti - where every module is a top level module... without proper modularisation.

I'd remove this change from this PR and I'm all for considering another PR that really tackles the problem here, providing a good alternative that still promotes modularity. Deprecating this interface does not provide a better/safer solution, but instead forces to flatten all modules across the board, which is really bad IMO.